### ISSUE TITLE
libpqxx fails for multiprocess environment

### ISSUE THREAD

**Issue by anikulkarni** [jtv/libpqxx#426]
This is a bit weird issue we are seeing. As such, the libpqxx works great and we had no issues so far. Recently, we added forking a child process periodically in our product code, which does some storage commands ('zfs' commands) and the main process (where libpq/xx is loaded) does IPC to this child process using a pipe. The child process and cleanup and IPC etc. work fine. But since the induction of this new code, we started getting the 'broken_connection' from libpqxx, which is the underlying libpq just closes the descriptor causing error returned as 'could not receive data from server: Bad file descriptor.'

The libpqxx use is straightforward, and we know that is not an issue as it was fine till IPC was inducted in the code. Here's the code snippet for generic command execution that we use:

```
//process.h
#include <string>
#include <vector>
#include <map>
#include <unistd.h>
#include <stdlib.h>

/////////////////////////////////////////////////////////////////////
namespace rw { namespace io {

/////////////////////////////////////////////////////////////////////
class Process
{
public:
        Process( std::string const& cmd, std::vector<std::string> const& args );
        ~Process();

        std::string GetLastError() const { return m_lastError; }

        // Start/stop process (stop will be kill of process-group)
        bool Start();
        bool Kill( int signal );
        bool IsAlive();
        bool Wait();
        int  GetPID()      const { return m_pid;    }
        int  GetExitCode() const { return m_status; }

        void SetEnvironment( std::map<std::string,std::string> const& envVarMap );

        // Interact with the process terminal
        bool ReadStdout( uint64_t length, std::string& output       );
        bool ReadStderr( uint64_t length, std::string& output       );
        bool TimedReadStdout( uint64_t length, uint64_t timeoutSecs, std::string& output );
        bool TimedReadStderr( uint64_t length, uint64_t timeoutSecs, std::string& output );
        bool WriteStdin( std::string const& output                  );

private:
        std::string                       m_cmd;
        std::vector<std::string>          m_args;
        int                               m_stdinPipe [2];
        int                               m_stdoutPipe[2];
        int                               m_stderrPipe[2];
        std::string                       m_lastError;
        bool                              m_running;
        pid_t                             m_pid;
        int                               m_status;
};

} } // namespace rw::io
``` 

```
//process.cpp
#include <rw/process.h>
#include <sstream>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <signal.h>
#include <errno.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <stdlib.h>

/////////////////////////////////////////////////////////////////////
using namespace std;
using namespace rw::io;


/////////////////////////////////////////////////////////////////////
Process::Process( string const& cmd, vector<string> const& args )
        : m_cmd( cmd ), m_args( args ), m_running( false ), m_pid( 0 ), m_status( 0 )
{
}

Process::~Process()
{
        if( IsAlive() )
        {
                Kill( SIGTERM );
                Wait();
        }

        close( m_stdinPipe[0] );
        close( m_stdinPipe[1] );

        close( m_stdoutPipe[0] );
        close( m_stdoutPipe[1] );

        close( m_stderrPipe[0] );
        close( m_stderrPipe[1] );
}

/////////////////////////////////////////////////////////////////////
// Start/stop process (stop will be kill of process-group)
bool Process::Start()
{
        char* args[ m_args.size() + 2 ];

        args[0] = (char *)m_cmd.c_str();

        for( uint64_t i=0; i < m_args.size(); ++i )
                args[i+1] = (char *)m_args[i].c_str();

        args[m_args.size() + 1] = (char *)0;

        pipe( m_stdinPipe  );
        pipe( m_stdoutPipe );
        pipe( m_stderrPipe );

        // Set our (parent) pipe end as non-blocking for stdout/stderr read
        fcntl( m_stdoutPipe[0], F_SETFD, O_NONBLOCK );
        fcntl( m_stderrPipe[0], F_SETFD, O_NONBLOCK );

        m_pid = fork();

        if( m_pid == 0 ) // New child process
        {
                close( m_stdinPipe [1] ); // Close write end
                close( m_stdoutPipe[0] ); // Close read end
                close( m_stderrPipe[0] ); // Close read end

                dup2( m_stdinPipe [0], 0 ); // Set new stdin
                dup2( m_stdoutPipe[1], 1 ); // Set new stdout
                dup2( m_stderrPipe[1], 2 ); // Set new stderr

                if( execvp( m_cmd.c_str(), args ) < 0 )
                {
                        m_lastError = "Failed to run child process: " + m_cmd;
                        exit( errno );
                }
        }
        else if( m_pid > 0 ) // Parent process
        {
                m_running = true;

                close( m_stdinPipe [0] ); // Close read end
                close( m_stdoutPipe[1] ); // Close write end
                close( m_stderrPipe[1] ); // Close write end

                return true;
        }
        else // Error - still in parent
        {
                m_lastError = "Failed to fork out child process: ";
                m_lastError+= strerror( errno );
                return false;
        }

        return false;
}

bool Process::Kill( int signal )
{
        if( !IsAlive() )
        {
                m_lastError = "Process is not running.";
                return false;
        }

        pid_t childPGID = getpgid( m_pid );

        if( childPGID < 0 )
        {
                m_lastError = "Failed to get pgid for the child process: ";
                m_lastError+= strerror( errno );
                return false;
        }

        pid_t callerPGID = getpgid( 0 );

        if( callerPGID != childPGID )
        {
                if( killpg( childPGID, signal ) != 0 )
                {
                        m_lastError = "Failed to kill the process: ";
                        m_lastError+= strerror( errno );
                        return false;
                }
        }
        else
        {
                if( kill( m_pid, signal ) != 0 )
                {
                        m_lastError = "Failed to kill the process: ";
                        m_lastError+= strerror( errno );
                        return false;
                }
        }

        m_running = false;

        return true;
}

bool Process::IsAlive()
{
        if( !m_running ) return false;


        pid_t rc = waitpid( m_pid, &m_status, WNOHANG );

        if( rc < 0 )
        {
                m_lastError = "Failed to wait for child process pid: ";
                m_lastError+= strerror( errno );
                return false;
        }
        else if( rc == 0 )
        {
                // process is still running
                return true;
        }
        else if( rc == m_pid )
        {
                // process finished
                m_running = false;
        }

        return m_running;
}

bool Process::Wait()
{
        if( m_pid == 0 )
        {
                m_lastError = "Process was never started.";
                return false;
        }

        int exitCode = 0;

        stringstream ss;

        if( m_running )
        {
                pid_t rc = waitpid( m_pid, &m_status, 0 );
                if( rc == -1 && errno != ECHILD )
                {
                        m_lastError = "Failed to wait for child process pid: ";
                        m_lastError+= strerror( errno );
                        return false;
                }
        }

        if( WIFEXITED( m_status ) )
        {
                exitCode = WEXITSTATUS( m_status );
                ss << "The process exited with code " <<  exitCode;
        }
        else if( WIFSIGNALED( m_status ) )
        {
                exitCode = WTERMSIG( m_status );
                ss << "The process killed by a signal " << exitCode;
        }
        else if( WIFSTOPPED( m_status ) )
        {
                exitCode = WSTOPSIG( m_status );
                ss << "The process stopped by a signal " << exitCode;
        }

        m_lastError = ss.str();

        m_status = exitCode;

        m_running = false;

        return true;
}

/////////////////////////////////////////////////////////////////////
void Process::SetEnvironment( map<string,string> const & envVarMap )
{
        for( auto const&[envVar,value] : envVarMap )
        {
                setenv( envVar.c_str(), value.c_str(),  /*overwrite*/1 );
        }
}

/////////////////////////////////////////////////////////////////////
// Interact with the process terminal
// Zero timeout is essentially infinite wait (blocking IO).
bool ReadTerminal( int fd, uint64_t length, uint64_t timeoutSecs, string& output, string& errMsg )
{
        if( length == 0 ) return true;

        char buf[length];
        memset( buf, 0, sizeof(buf) );

        uint64_t waitSecs = 0;

        int rc = 0;
        while( waitSecs <= timeoutSecs || timeoutSecs == 0 )
        {
                rc = read( fd, buf, sizeof(buf)-1 );

                if( rc == EAGAIN )
                {
                        sleep(1);
                        waitSecs += 1;
                        continue;
                }
                else if( rc < 0 )
                {
                        errMsg = "Failed to read a line from the pipe: ";
                        errMsg+= strerror( errno );
                        break;
                }

                buf[sizeof(buf)-1] = 0;
                output.assign( buf );
                break;
        }

        return rc == 0;
}

bool Process::ReadStdout( uint64_t length, string& output )
{
        return ReadTerminal( m_stdoutPipe[0], length, /*timeoutSecs*/ 0, output, m_lastError );
}

bool Process::ReadStderr( uint64_t length, string& output )
{
        return ReadTerminal( m_stderrPipe[0], length, /*timeoutSecs*/ 0, output, m_lastError );
}

bool Process::TimedReadStdout( uint64_t length, uint64_t timeoutSecs, string& output )
{
        return ReadTerminal( m_stdoutPipe[0], length, timeoutSecs, output, m_lastError );
}

bool Process::TimedReadStderr( uint64_t length, uint64_t timeoutSecs, string& output )
{
        return ReadTerminal( m_stderrPipe[0], length, timeoutSecs, output, m_lastError );
}

bool Process::WriteStdin( string const& input )
{
        int rc = write( m_stdinPipe[1], input.c_str(), input.size() );

        if( rc < 0 )
        {
                m_lastError = "Failed to write a line to the pipe: ";
                m_lastError+= strerror( errno );
        }

        return rc == 0;
}

/////////////////////////////////////////////////////////////////////

```

(The above code is fine, but I am sharing to repro the issue easily. If you use it in a program using libpqxx, you can easily break the connection when the above fork+exec call happens).

We use TCP sockets for DB connection. After digging into libpq code, I found that as soon as the execve call happens, DB descriptor in libpq is closed (becomes invalid), so next time any libpqxx (and so libpq) API is called that tries to use that descriptor then the error (an exception in libpqxx) is generated. The error makes sense, as it is just telling that libpq tried to use the socket descriptor which is no more valid, but it is not clear why the descriptor is getting closed. It wouldn't make sense if the libpq is not handling typical IPC cases.

I checked libpq code and I saw that it correctly handles SIGCHLD interrupts (or potential other interrupts) while waiting for socket and retries correctly (when any socket call returns < 0 value with errno set to EINTR). I added some similar changes to libpqxx code too (like in wait_fd function) but no luck so far. The parent process where libpq/xx is loaded and where the above Process class is used already uses 'signal(SIGHLD, SIG_IGN)'. Also, I saw libpq employs FD_CLOEXEC (i.e., descriptor close-on-exec) everywhere it creates a new socket/descriptor, which is the right thing to do to avoid the descriptor leaks.

Any clues what could be wrong?  If I comment out the usage of the Process class above (so IPC is no more used), then libpqxx/libpq work correctly, but I can't avoid IPC in my case. I am new to libpq code but due to this issue, I dug into that code to a good extent. I think it is doing some counting for descriptor usage on its own which might be an issue here. But would appreciate any clues. Is this something potentially caused by libpqxx?

**Comment by jtv** [comment]
I'm having to read this on my phone, so I can't really look at the code.
It is a lot of code, so I doubt I'll get around to reading it fully.
Please always simplify it to the absolute minimal example that reproduces
the problem!  Often that step alone will reveal the cause of the problem.

The one thing I'm wondering is: could there be a problem with `fork()`
duplicating the socket inside the connection?  I could imagine a few things
going wrong, such as: you'd probably get a mess if a transaction was active
at the time of the fork.

**Comment by anikulkarni** [comment]
The example is a simple Process class, which is a wrapper to execute any command. The reason I put the entire example is so it is compilable and can be used in a sample code to reproduce the issue easily (you can run any command with it). Maybe since you are reading on phone, the example may look complicated/big but it's really straightforward.



> The one thing I'm wondering is: could there be a problem with `fork()`
> duplicating the socket inside the connection?  I could imagine a few things
> going wrong, such as: you'd probably get a mess if a transaction was active
> at the time of the fork.

Your guess is as good as mine, so that is the place I started looking at (The one more reason I posted wrapper class code doing the fork). I could not find anything unusual for the rw::io::Process class wrapper that does a fork. It simply calls fork, then attaches child end of the pipes (pipes are from the parent process) to stdout/stderr/stdin resp, so any command run later with execve call would be readable for stdout/err from the parent process and pipe created for IPC. The fork and all the logic is fairly common as far as I see. That is what the puzzling part for me atm, as it is all usual fork business and it is not clear why libpq/xx is duplicating the descriptor to child process even though close-on-exec is set for the DB connection socket (I tried adding as well as removing close-on-exec for all descriptors around socket in libpq and nothing helped). Moreover, it is not clear how libpq/xx's socket duplicated to child process causes parent process socket to get closed, as per POSIX documentation, descriptors in child process are cloned, so shouldn't cause the issue to a descriptor in the parent process (irrespective of what child process does with it).

At this stage, I am stumped to what causes descriptor to get closed. Also, I see this in the postgres log:
```
2021-03-20 05:28:55.412 EDT [14770] LOG:  unexpected EOF on client connection with an open transaction
2021-03-20 05:29:24.425 EDT [14806] LOG:  could not receive data from client: Connection reset by peer
2021-03-20 05:29:24.433 EDT [14811] LOG:  could not receive data from client: Connection reset by peer
2021-03-20 05:29:24.437 EDT [14815] LOG:  could not receive data from client: Connection reset by peer
```

so it is clear that bad descriptor is indeed libpq/xx abruptly closing the socket the moment fork+execve completes.


I am hoping someone here uses libpq/xx in an environment where IPC is done, so can shed some light to potential areas at least where I can look next.

**Comment by jrch2k20** [comment]
hey, anikul i had this issue as well but with threads and/or async lambdas because basically pqxx is not thread friendly (note mongo  and redis libraries have the same issues).

My solution was to simply pass around the connection string instead of the actual object and recreate the connection on the actual places are needed hence the objects remain local to that scope they are called(work for mongo and redis as well). Note you can pass that object around as long as it remains on the same process/thread just fine.

As for performance, i really couldn't measure any meaningful impact and stability skyrocketed

**Comment by jtv** [comment]
@anikulkarni could it be simply that the child closes its fd and renders the fd in the parent unusable, messing up the destructor there?  I think @jrch2k20 has a point about passing connection strings — I wouldn't even say that libpq and libpqxx aren't friendly to threads, just that passing complex mutable state between threads of execution is inherently risky.

**Comment by KayEss** [comment]
Ideally you'd fork only near the point in time that your program starts -- certainly before you start any threads and ideally before doing anything too interesting. The epoll APIs are completely broken with forks (but I haven't seen any evidence that either libpq or libpqxx use epoll).

What I've done in the past is to fork on start up and use a pipe to have the main executable tell it's child whenever it needs to fork and then the child does the fork instead of the parent.  This is totally reliable and side steps any and all weirdness.

**Comment by anikulkarni** [comment]
Actually, the libpq uses select/epoll/poll and I found that as I was seeing another issue (As a side effect of the above bug) wherein the libpq would wait on the DB socket forever - It wouldn't detect DB socket closed.

The Process class above is a simple utility class doing fork business (Can call any system commands, and we do 'ZFS' commands). I assume ZFS commands close all descriptors, but the mystery part right now is why the DB socket (and any other descriptors) leak to the child process, as 
1. All descriptors managed by libpq are opened with close-on-exec (FD_CLOEXEC set on it at the right time and place). I also added similar missing checks for one descriptor I found in the libpqxx. The CLOEXEC (close-on-exec) should make sure that the moment rw::io::Process class above did a child process (with fork and exec), all descriptors cloned into child are closed gracefully, at least per POSIX documentation for the flag.
2.  The rw::io::Process code above is invoked is in different thread than the thread (or thread-pool) doing DB connections with libpqxx, though they are part of the same process which does child processes with the Process class pasted earlier (that does fork+execve).

It is clear that somehow, libpq created descriptors are closed when you do a child process (with a fork) and do execXX call there. It doesn't matter where your fork-and-exec flow happens (Could be in an irrelevant place to all DB management). To repro, you can simply add a program doing DB calls with the above Process class code (Just instantiate it and run any random Linux command execution through the Process class). You will find that the DB connection breaks/closes. I suspected Process class code as the culprit first but I could repro the issue with replacing that code with the std::system() API too.

Anyway, for now, given my findings, and in general what you all are suggesting too, I reorganized my program code to move all fork-and-exec business to a separate process, which exposes webservices (through libmicrohttpd) to the main process. This avoids the issue. The reorg was actually okay in terms of design as the new webservice can be used on remote hosts too now. But I am sure people must be seeing this issue, as using libpq/libpqxx in a multiprocess environment (with IPC) must be a too common scenario.

**Comment by KayEss** [comment]
epoll in particular suffers a design flaw in that closing the FD in the child also closes it in the parent. In order to survive the fork the FDs have to be removed from epoll before the fork. This is why ASIO has to have [this forking nonsense](https://think-async.com/Asio/asio-1.18.1/doc/asio/reference/io_context/notify_fork.html) and it looks like something similar may also leak into the networking TS and a future C++ standard -- hopefully we come to our senses before that happens.
